components/script/dom/webgl2renderingcontext.rs

```diff
@@ -1524,6 +1707,53 @@ impl WebGL2RenderingContextMethods for WebGL2RenderingContext {
         self.base.Uniform4iv(location, v)
     }
 
+    // https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.8
+    fn Uniform4ui(&self, location: Option<&WebGLUniformLocation>, x: u32, y: u32, z: u32, w: u32) {
+        self.base.with_location(location, |location| {
+            match location.type_() {
+                constants::BOOL_VEC4 | constants::UNSIGNED_INT_VEC4 => {},
+                _ => return Err(InvalidOperation),
+            }
+            self.base
+                .send_command(WebGLCommand::Uniform4ui(location.id(), x, y, z, w));
+            Ok(())
+        });
+    }
+
+    // https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.8
+    fn Uniform4uiv(
+        &self,
+        location: Option<&WebGLUniformLocation>,
+        val: Uint32ArrayOrUnsignedLongSequence,
+        src_offset: u32,
+        src_length: u32,
+    ) {
+        self.base.with_location(location, |location| {
+            match location.type_() {
+                constants::BOOL_VEC4 | constants::UNSIGNED_INT_VEC4 => {},
+                _ => return Err(InvalidOperation),
+            }
+            let val = match val {
+                Uint32ArrayOrUnsignedLongSequence::Uint32Array(v) => v.to_vec(),
+                Uint32ArrayOrUnsignedLongSequence::UnsignedLongSequence(v) => v,
+            };
+
+            let (offset, length) =
+                self.check_uniform_vec_section(src_offset, src_length, val.len())?;
+            let val = val[offset..offset + length].to_vec();
+            if val.len() < 4 || val.len() % 4 != 0 {
```

Comment:
Why don't we change check_uniform_vec_section to reduce some code duplication in these methods?
* it can take a Uint32ArrayOrUnsignedLongSequence argument and extract the vector
* it can check if offset is 0 and length is the vector length and return the original vector without copying anything
* it can take an argument for uniform size and return an error if the length doesn't match
* it can take a location argument and return an error if the location size is none and the array length doesn't match
* finally it can return the adjusted vector

Reply:
Ok, updated, thanks!